src/core/evaluator.js

```diff
@@ -702,10 +703,19 @@ class PartialEvaluator {
 
     const softMask = dict.get("SM", "SMask") || false;
     const mask = dict.get("Mask") || false;
+    const canUseOffscreenCanvas =
+      this.options.isOffscreenCanvasSupported &&
+      FeatureTest.isOffscreenCanvasSupported;
 
     const SMALL_IMAGE_DIMENSIONS = 200;
     // Inlining small images into the queue as RGB data
-    if (isInline && !softMask && !mask && w + h < SMALL_IMAGE_DIMENSIONS) {
+    if (
+      !canUseOffscreenCanvas &&
```

Comment:
How does this affect overall performance, since I believe that the idea was that *synchronously* decoding these very small/simple inline images was faster than the "general" (and async) image parsing below?

Furthermore, in the `src/core/operator_list.js` file there's optimizations for sequences of inline images and it's not all clear to me (from a glance) how these changes would affect performance for PDF documents that benefit from that code!?

Reply:
Without this patch, when `forceRGBA` is true we start to create the image here:
https://github.com/mozilla/pdf.js/blob/546902df63b4fcfad419bc8109d128040d7b27ab/src/core/image.js#L751
So for a small image we just decode it in RGBA format in the worker and then it can be drawn like this (modulo transfer maps) in the main thread: it's faster and doesn't use that much extra memory.

With this patch: all the images are decoded in RGBA in the worker and then converted into a Bitmap. The rendering in the main thread just takes less than few millisecs and it avoid to block the main thread. Consequently, I'd say that there are no possible regressions here (but I can be wrong).

In term of memory, before the patch we've a buffer for the image (1BPP or RGB or RGBA) and we create a canvas for it. The cache is containing the buffer.
With this patch, we still create an OffscreenCanvas for the image and we don't have the buffer anymore but the cache contains the bitmap.

If you could find some pdfs which could be problematic because of this patch, please share them and I'll see if we can do something better.

Reply:
Sorry, but I don't think this really addresses the question about the now (essentially) unused optimizations in the `src/core/operator_list.js` file; I'm referring in particular to this code: https://github.com/mozilla/pdf.js/blob/13af3f3442d1e6c0b08c39a54605f6bd5b564b1c/src/core/operator_list.js#L39-L44

A quick look at `blame` points to PR #2417, which seems to contain references to test-cases.